### PR TITLE
Allow using custom SSL certificates

### LINKED ISSUE #8126: Allow other certificates
When working behind a SSL proxy, the SSL proxy presents another SSL certificate. When not using that certificate, the DOI fetcher returns "Connection error" without any further information:

image

IMHO JabRef should offer the possibility to:

- use an own key store
  - activated through preferences
- import `.cer` files into that key store
  - possibility offered using the preferences

See https://stackoverflow.com/q/18889058/873282 for some code hints.

Somehow refs https://github.com/JabRef/jabref/issues/7627. Here, we have an SSL certificate issue. There, we have a proxy authentication error.

### PR DESCRIPTION
**Pull request by HoussemNasri**
Fixes #8126.
### Context
When a browser or server attempts to connect to a website (i.e., a web server) secured with SSL, the browser requests that the web server identifies itself by sending a copy of its SSL certificate in response then it checks to see whether it trusts the SSL certificate. If it does, it signals this to the webserver and starts an SSL encrypted session; if it doesn't the connection fails. The Java HTTP client uses a similar method.

**So, how does Java trust certificates?**
The JVM stores a list of trusted certificates in a truststore. Normally Java uses a system-wide truststore:
- **Java 8:** *$JAVA_HOME/jre/lib/security/cacerts*
- **Java 11:** *$JAVA_HOME/lib/security/cacerts*

However it is possible to use a different truststore by specifying a parameter, `-Djavax.net.ssl.trustStore=/path/to/truststore`, where `/path/to/truststore` is the absolute file path of the alternative truststore.

When connecting to a server over HTTPS, the Java client looks up the associated certificate in our truststore. If the certificate or Certificate Authorities presented by the external server is not in our truststore, we'll get an *SSLHandshakeException* and the connection won't be set up successfully. 

### Problem Description
Using an SSL proxy, the Java client requests a certificate from a server through the SSL proxy, which intercepts the request and provides its own SSL certificate, which may or may not be available in our truststore

### Solution
The solution is to copy the original JDK truststore into the project and edit it, such as adding or removing certificates based on user interactions. It also involves persisting user-created certificates for the UI.

#### UI Changes
I added a new section to network preferences for SSL configurations, this section contains a table of user-added certificates, with basic information about each certificate, such as *serial number*, *version*, *validation dates* and more.
#### Screenshots
Screenshot 2022-03-26 140542

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by HoussemNasri** [comment]
I tried putting the truststore file in resources, but it doesn't seem to work; the certificates don't seem to persist. I think I should put it in application data, but I'm not sure how to do that across operating systems.

**Comment by Siedlerchr** [comment]
For the app data dir  thing, checkout the PdfIndexer class, we use an external library which uses an external library that provides OS independent path to app data

**Comment by Siedlerchr** [comment]
Alternative idea: Ask the user to store the file somewhere

**Comment by HoussemNasri** [comment]
We can add another setting to SSL configuration to choose the truststore path; I'll think about it later once I've set up SSL configuration preference storing logic; then it'll be a simple task.

**Comment by HoussemNasri** [comment]
I've set up an HTTPS server on my machine with a self-signed certificate and it worked :)

File: src/main/java/org/jabref/gui/preferences/network/NetworkTab.java [hunk 1]
```diff
@@ -102,6 +118,28 @@ public void initialize() {
...
+        certSignatureAlgorithm.setCellValueFactory(data -> data.getValue().signatureAlgorithmProperty());
+        certVersion.setCellValueFactory(data -> EasyBind.map(data.getValue().versionProperty(), this::formatVersion));
+
+        certValidFrom.setCellValueFactory(data -> EasyBind.map(data.getValue().validFromProperty(), this::formatDate));
+        certValidTo.setCellValueFactory(data -> EasyBind.map(data.getValue().validToProperty(), this::formatDate));
+
+        customCertificatesTable.itemsProperty().set(viewModel.customCertificateListProperty());
+    }
+
+    public String formatDate(LocalDate localDate) {
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/preferences/network/NetworkTab.java:137]: private?

File: src/main/java/org/jabref/gui/preferences/network/NetworkTabViewModel.java [hunk 2]
```diff
@@ -179,6 +230,30 @@ private void storeProxySettings(ProxyPreferences newProxyPreferences) {
         proxyPreferences.setPassword(newProxyPreferences.getPassword());
     }
 
+    public void storeSSLSettings() {
+        trustStoreManager.flush();
+        sslPreferences.setUseCustomCertificates(customCertificatesUseProperty.getValue());
+        sslPreferences.setCustomCertificateVersion(customCertificateListProperty.stream()
+                                                                                .map(CustomCertificateViewModel::getVersion)
+                                                                                .collect(Collectors.toList()));
+        sslPreferences.setCustomCertificateThumbprint(customCertificateListProperty.stream().map(CustomCertificateViewModel::getThumbprint).collect(Collectors.toList()));
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/preferences/network/NetworkTabViewModel.java:239]: newLine

File: src/main/java/org/jabref/gui/preferences/network/NetworkTabViewModel.java [hunk 3]
```diff
@@ -179,6 +230,30 @@ private void storeProxySettings(ProxyPreferences newProxyPreferences) {
...
+                                                                                .map(this::localDateToEpochDayStr)
+                                                                                .collect(Collectors.toList()));
+        sslPreferences.setCustomCertificateSerialNumber(customCertificateListProperty.stream().map(CustomCertificateViewModel::getSerialNumber).collect(Collectors.toList()));
+        sslPreferences.setCustomCertificateIssuer(customCertificateListProperty.stream().map(CustomCertificateViewModel::getIssuer).collect(Collectors.toList()));
+        sslPreferences.setCustomCertificateSigAlgorithm(customCertificateListProperty.stream().map(CustomCertificateViewModel::getSignatureAlgorithm).collect(Collectors.toList()));
+    }
+
+    private String localDateToEpochDayStr(LocalDate date) {
+        return String.valueOf(date.toEpochDay());
+    }
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/preferences/network/NetworkTabViewModel.java:255]: refactoring to a separate method produces more boilerplate than inlining it. Or is there a different reason for extracting this?

File: src/main/java/org/jabref/gui/preferences/network/NetworkTabViewModel.java [hunk 4]
```diff
@@ -309,4 +382,32 @@ public StringProperty proxyUsernameProperty() {
...
+                .build();
+
+        dialogService.showFileOpenDialog(fileDialogConfiguration).ifPresent(certPath -> {
+            SSLCertificate.fromPath(certPath).ifPresent(sslCertificate -> {
+                if (!trustStoreManager.isCertificateExist(sslCertificate.getSHA256Thumbprint())) {
+                    trustStoreManager.addCertificate(sslCertificate.getSHA256Thumbprint(), certPath);
+
+                    customCertificateListProperty.add(CustomCertificateViewModel.fromSSLCertificate(sslCertificate));
+                } else {
+                    // TODO('Show a dialog or toast message indicating that the user is trying to add a duplicate certificate')
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/preferences/network/NetworkTabViewModel.java:408]: dialogService.showDialogAndWait?

File: src/main/java/org/jabref/logic/net/URLDownload.java [hunk 5]
```diff
@@ -99,7 +98,9 @@ public URLDownload(URL source) {
      * security-relevant information this is kind of OK (no, actually it is not...).
      * <p>
      * Taken from http://stackoverflow.com/a/6055903/873661 and https://stackoverflow.com/a/19542614/873661
+     * @deprecated
      */
+    @Deprecated
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/net/URLDownload.java:103]: 🎉

File: src/main/java/org/jabref/preferences/JabRefPreferences.java [hunk 6]
```diff
@@ -1604,6 +1637,43 @@ public ProxyPreferences getProxyPreferences() {
...
+                getBoolean(CUSTOM_CERTIFICATE_USE),
+                getStringList(CUSTOM_CERTIFICATE_THUMBPRINT),
+                getStringList(CUSTOM_CERTIFICATE_VERSION),
+                getStringList(CUSTOM_CERTIFICATE_VALID_FROM),
+                getStringList(CUSTOM_CERTIFICATE_VALID_TO),
+                getStringList(CUSTOM_CERTIFICATE_SERIAL_NUMBER),
+                getStringList(CUSTOM_CERTIFICATE_ISSUER),
+                getStringList(CUSTOM_CERTIFICATE_SIG_ALGORITHM),
+                get(TRUSTSTORE_PATH)
+        );
```
Comment (calixtus) [inline src/main/java/org/jabref/preferences/JabRefPreferences.java:1656]: Im guessing you read them as string lists to be able to observe them. There is maybe a solution to parse the certificates already here and to create an array of observables for every certificate. See BibEntry::getObservables . Would that be a cleaner solution?

**Review by calixtus — COMMENTED** [review]
On first view it looks really cool! I like it very much.
However: checkstyle complains and i've got some little nitpicks and a question.
Thanks!

File: src/main/java/org/jabref/preferences/JabRefPreferences.java (hunk 6 above)
Comment (HoussemNasri) [reply src/main/java/org/jabref/preferences/JabRefPreferences.java]: Speaking about it made me realize that we don't need to store that information anywhere else; it's already in the truststore. The disadvantage is that we can't store custom properties; we're stuck with what the `X509Certificate` class provides. I updated the code to use the truststore as the single source of truth, I think this is better, the alternative brings lots of syncing issues

File: src/main/java/org/jabref/gui/preferences/network/NetworkTabViewModel.java (hunk 4 above)
Comment (HoussemNasri) [reply src/main/java/org/jabref/gui/preferences/network/NetworkTabViewModel.java]: Done

File: src/main/java/org/jabref/gui/preferences/network/NetworkTab.java (hunk 1 above)
Comment (HoussemNasri) [reply src/main/java/org/jabref/gui/preferences/network/NetworkTab.java]: Done

File: src/main/java/org/jabref/logic/net/ssl/TrustStoreManager.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+    private final Path storePath;
+
+    private KeyStore store;
+
+    public TrustStoreManager(Path storePath) {
+        this.storePath = storePath;
+        try {
+            LOGGER.info("Trust store path: {}", storePath.toAbsolutePath());
+            Path storeResourcePath = Path.of(TrustStoreManager.class.getResource("/ssl/truststore.jks").toURI());
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/net/ssl/TrustStoreManager.java:36]: This definitely needs to be checked at runime in  a jpackage build.

**Comment by Siedlerchr** [comment]
Codewise lgtm! Needs to be tested when run from jpackage to be sure that all resources paths can be accessed.
Can you test it with ./gradlew jpackage ? 
That should generate the binaries under build/distribution

**Review by Siedlerchr — APPROVED** [review]

**Comment by HoussemNasri** [comment]
I tested the installer and portable versions on windows, and they both work.

**Comment by calixtus** [comment]
Please fixe the language test, so we can merge.

**Review by Siedlerchr — APPROVED** [review]
Thanks a lot for this cool feature!

**Pull request merged**